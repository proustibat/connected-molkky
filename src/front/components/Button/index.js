import PropTypes from 'prop-types';
import React from 'react';

const Button = ({
  onClick, children, disabled, size, style, className,
}) => (
  <button
    type="button"
    className={`waves-effect waves-light z-depth-1 btn ${size}${disabled ? ' disabled' : ''}${className ? ` ${className}` : ''}`}
    onClick={onClick}
    style={style}
  >
    {children}
  </button>
);


Button.propTypes = {
  onClick: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
  disabled: PropTypes.bool,
  size: PropTypes.oneOf(['btn-large', 'btn-small', '']),
  style: PropTypes.object,
  className: PropTypes.string,
};

Button.defaultProps = {
  disabled: false,
  size: 'btn-large',
  style: {},
  className: '',
};

Button.displayName = 'Button';

export default Button;
