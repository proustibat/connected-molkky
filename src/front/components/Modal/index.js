import React, { useEffect, useLayoutEffect, useState } from 'react';
import PropTypes from 'prop-types';
import hasIn from 'lodash/hasIn';

const Modal = ({
  id, children, onEditModalInit, title, footer, onCloseEnd,
}) => {
  const [instance, setInstance] = useState(null);

  useLayoutEffect(() => {
    if (hasIn(global, 'M.Modal.init')) {
      const modalInstance = global.M.Modal.init(document.querySelector(`#${id}`), {
        onCloseEnd,
        dismissible: false,
      });
      setInstance(modalInstance);
    }
    return () => {
      if (hasIn(global, 'M.Modal.getInstance')) {
        const modalInstance = global.M.Modal.getInstance(document.querySelector(`#${id}`));
        modalInstance && modalInstance.destroy();
      }
    };
  }, []);

  useEffect(() => {
    instance && onEditModalInit(instance);
  }, [instance]);

  return (
    <div id={id} className="modal modal-fixed-footer">
      <div className="modal-content">
        {title && <h2 style={{ borderLeft: '1rem solid', borderRight: '1rem solid' }}>{title}</h2>}
        {children}
      </div>
      <div className="modal-footer">
        {footer && footer}
      </div>
    </div>
  );
};

Modal.propTypes = {
  id: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  onEditModalInit: PropTypes.func.isRequired,
  onCloseEnd: PropTypes.func.isRequired,
  title: PropTypes.string,
  footer: PropTypes.node,
};

Modal.defaultProps = {
  title: null,
  footer: null,
};

export default Modal;
