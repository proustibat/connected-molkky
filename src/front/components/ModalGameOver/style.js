export default {
  modalFooterButton: { margin: 0 },
  modalContent: {
    display: 'flex',
    flexDirection: 'column',
    height: 'calc(100% - 43px)',
    padding: 0,
    margin: 0,
    justifyContent: 'space-evenly',
  },
  winnerName: { textAlign: 'center', margin: 0 },
  winnerIconsContainer: { position: 'relative' },
  trophy: {
    width: '15rem',
    height: '15rem',
    display: 'block',
    margin: 'auto',
    fill: '#143746',
  },
  winnerIcon: {
    width: '3.5rem',
    marginRight: '0.6rem',
    marginLeft: '0.6rem',
    position: 'absolute',
    top: '5rem',
    left: 'calc(50% - 2.25rem)',
  },
  winnerScore: {
    position: 'absolute',
    top: 0,
    left: 'calc(50% - 2rem)',
    color: '#e2e2e2',
    fontSize: '2rem',
    textAlign: 'center',
    width: '4rem',
  },
  looserContainer: { textAlign: 'center', margin: 0 },
  loserIcon: { width: '2rem', verticalAlign: 'middle' },
};
