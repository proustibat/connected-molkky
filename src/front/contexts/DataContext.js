import React, { useContext } from 'react';

export const DataContext = React.createContext([]);
export const useDataContext = () => useContext(DataContext);
export const DataContextProvider = DataContext.Provider;
export const DataContextConsumer = DataContext.Consumer;

export default DataContext;
