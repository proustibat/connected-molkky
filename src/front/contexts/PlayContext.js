import React from 'react';

export const PlayContext = React.createContext({});
export const usePlayContext = () => React.useContext(PlayContext);
export const PlayContextProvider = PlayContext.Provider;
export const PlayContextConsumer = PlayContext.Consumer;

export default PlayContext;
