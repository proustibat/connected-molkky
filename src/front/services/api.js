import { toastError } from '@root/front/services/toasts';

// eslint-disable-next-line import/prefer-default-export
export const startGame = ({ teams, playingTeam }) => fetch('/api/molkky/start', {
  method: 'post',
  headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
  body: JSON.stringify({ teams, playingTeam }),
})
  .then((response) => (response.status === 200
    ? response.json()
    : Promise.reject(response.statusText)))
  .catch(toastError);

export const scorePoints = ({ team, points }) => fetch('/api/molkky/score', {
  method: 'post',
  headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
  body: JSON.stringify({ team, points }),
})
  .then((response) => (response.status === 200
    ? response.json()
    : Promise.reject(response.statusText)))
  .catch(toastError);

export const missTarget = (team) => fetch('/api/molkky/miss', {
  method: 'post',
  headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
  body: JSON.stringify({ team }),
})
  .then((response) => (response.status === 200
    ? response.json()
    : Promise.reject(response.statusText)))
  .catch(toastError);

export const resetGame = () => fetch('/api/molkky/reset', {
  method: 'post',
  headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
})
  .then((response) => (response.status === 200
    ? response.json()
    : Promise.reject(response.statusText)))
  .catch(toastError);

export const startFakeSkittles = () => fetch('/api/mocks/skittles/start', {
  method: 'post',
  headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
})
  .then((response) => (response.status === 200
    ? response.json()
    : Promise.reject(response.statusText)))
  .catch(toastError);

export const stopFakeSkittles = () => fetch('/api/mocks/skittles/stop', {
  method: 'post',
  headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
})
  .then((response) => (response.status === 200
    ? response.json()
    : Promise.reject(response.statusText)))
  .catch(toastError);

export const simulateThrow = () => fetch('/api/mocks/skittles/simulate-throw', {
  method: 'post',
  headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
})
  .then((response) => (response.status === 200
    ? response.json()
    : Promise.reject(response.statusText)))
  .catch(toastError);
