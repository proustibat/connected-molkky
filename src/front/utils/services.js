import constants from '@utils/constants';

// eslint-disable-next-line import/prefer-default-export
export const getRandomPosition = () => {
  const randomValue = Math.floor(Math.random() * Math.floor(100));
  const conditions = [{
    predicate: (value) => value >= 3,
    position: constants.POSITION.UPRIGHT,
  }, {
    predicate: (value) => value < 3 && value !== 0,
    position: constants.POSITION.KNOCKED_OVER,
  }, {
    predicate: () => true,
    position: null,
  }];
  return conditions.find((condition) => condition.predicate(randomValue)).position;
};

export const getRandomPositionData = () => Array.from({ length: 12 }, (_, i) => ({
  value: i + 1,
  position: getRandomPosition(),
}));
