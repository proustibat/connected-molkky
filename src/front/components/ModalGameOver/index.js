import React, { useState } from 'react';
import Button from '@components/Button';
import Modal from '@components/Modal';
import PropTypes from 'prop-types';
import { resetGame } from '@root/front/services/api';
import { useDataContext } from '@contexts/DataContext';
import { useHistory } from 'react-router-dom';

const ModalGameOver = ({
  winnerName, winnerIcon: SVGWinnerIcon, winnerScore,
  loserName, loserIcon: SVGLoserIcon, loserScore,
  onInit,
}) => {
  const { isLoading, setIsLoading } = useDataContext();
  const history = useHistory();

  const [modalInstance, setModalInstance] = useState(null);

  const onPlayAgain = async () => {
    setIsLoading(true);
    const result = await resetGame();
    setIsLoading(false);
    if (result) {
      modalInstance.close();
    }
  };

  const onModalInit = (instance) => {
    setModalInstance(instance);
    onInit(instance);
  };

  const onModalCloseEnd = () => {
    history.push('/molkky/game');
  };

  return (
    <Modal
      id="game-over-modal"
      onInit={onModalInit}
      onCloseEnd={onModalCloseEnd}
      title="GAME OVER"
      styleTitle={{ justifyContent: 'center' }}
      footer={(
        <Button
          onClick={onPlayAgain}
          style={{ margin: 0 }}
          disabled={isLoading}
        >
          Play Again
        </Button>
      )}
    >
      <div style={{
        display: 'flex',
        flexDirection: 'column',
        height: 'calc(100% - 43px)',
        padding: 0,
        margin: 0,
        justifyContent: 'space-evenly',
      }}
      >
        <p className="flow-text" style={{ textAlign: 'center', margin: 0 }}>
          Boo-ya!
          <br />
          {winnerName}
          {' '}
          wins!
        </p>
        <div style={{ position: 'relative' }}>
          <svg
            style={{
              width: '15rem',
              height: '15rem',
              display: 'block',
              margin: 'auto',
              fill: '#143746',
            }}
            height="48"
            viewBox="0 0 48 48"
            width="48"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g data-name="winner" id="winner">
              <path
                style={{ fill: '#0c222b' }}
                d="M14.871,26.048C4.827,26.048.765,15.87.018,10.264A2,2,0,0,1,2,8H12a2,2,0,0,1,0,4H4.469c1.014,3.722,3.906,10.633,11.365,10.007a2,2,0,0,1,.332,3.986Q15.5,26.049,14.871,26.048Z"
              />
              <rect height="4" rx="2" ry="2" width="34" x="7" y="2" />
              <rect style={{ fill: '#0c222b' }} height="4" rx="2" ry="2" width="34" x="7" y="2" />
              <path d="M33.129,26.048q-.633,0-1.295-.055a2,2,0,1,1,.332-3.986c7.5.622,10.37-6.284,11.373-10.007H36a2,2,0,0,1,0-4H46a2,2,0,0,1,1.982,2.264C47.235,15.87,43.172,26.047,33.129,26.048Z" />
              <path style={{ fill: '#0c222b' }} d="M33.129,26.048q-.633,0-1.295-.055a2,2,0,1,1,.332-3.986c7.5.622,10.37-6.284,11.373-10.007H36a2,2,0,0,1,0-4H46a2,2,0,0,1,1.982,2.264C47.235,15.87,43.172,26.047,33.129,26.048Z" />
              <polygon style={{ fill: '#091c23' }} points="28 35 20 35 21 30 27 30 28 35" />
              <path style={{ fill: '#0c222b' }} d="M34,45H14V36a1.027,1.027,0,0,1,1.053-1H32.947A1.027,1.027,0,0,1,34,36Z" />
              <rect style={{ fill: '#ebebeb' }} height="5" width="12" x="18" y="38" />
              <path style={{ fill: '#143746' }} d="M10,6H38a0,0,0,0,1,0,0V17A14,14,0,0,1,24,31h0A14,14,0,0,1,10,17V6a0,0,0,0,1,0,0Z" />
              <rect style={{ fill: '#0c222b' }} height="3" rx="1" ry="1" width="28" x="10" y="45" />
            </g>
          </svg>
          <SVGWinnerIcon
            style={{
              width: '3.5rem',
              marginRight: '0.6rem',
              marginLeft: '0.6rem',
              position: 'absolute',
              top: '5rem',
              left: 'calc(50% - 2.25rem)',
            }}
          />
          <p style={{
            position: 'absolute',
            top: 0,
            left: 'calc(50% - 2rem)',
            color: '#e2e2e2',
            fontSize: '2rem',
            textAlign: 'center',
            width: '4rem',
          }}
          >
            {winnerScore}
          </p>
        </div>
        <p className="flow-text" style={{ textAlign: 'center', margin: 0 }}>
          Bwah-hah-hah!
          <br />
          <span role="img" aria-label="thumb down">&#128078;</span>
          <SVGLoserIcon style={{ width: '2rem', verticalAlign: 'middle' }} />
          <br />
          {loserName}
          {' '}
          loses with
          {' '}
          {loserScore}
          !
        </p>
      </div>
    </Modal>
  );
};

ModalGameOver.propTypes = {
  winnerName: PropTypes.string.isRequired,
  winnerIcon: PropTypes.func.isRequired,
  winnerScore: PropTypes.number.isRequired,
  loserName: PropTypes.string.isRequired,
  loserIcon: PropTypes.func.isRequired,
  loserScore: PropTypes.number.isRequired,
  onInit: PropTypes.func.isRequired,
};

export default ModalGameOver;
