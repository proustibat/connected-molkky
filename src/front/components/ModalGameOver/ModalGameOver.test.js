import * as Api from '@root/front/services/api';
import * as DataContextModule from '@contexts/DataContext';
import * as PlayContextModule from '@contexts/PlayContext';
import Button from '@components/Button';
import CatSVG from '@root/front/svg/cat.svg';
import DogSVG from '@root/front/svg/dog.svg';
import Modal from '@components/Modal';
import ModalGameOver from './index';
import React from 'react';
import { shallow } from 'enzyme';

jest.mock('react-router-dom', () => ({ useHistory: () => ({ push: jest.fn() }) }));

const givenProps = {
  winnerName: 'Team Cat',
  winnerIcon: CatSVG,
  winnerScore: 50,
  loserName: 'Team Dog',
  loserIcon: DogSVG,
  loserScore: 35,
};

const modalInstance = {};

describe('ModalGameOver', () => {
  let useDataContextSpy;
  let usePlayContextSpy;
  let resetGameSpy;

  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation((init) => [init, setState]);

  beforeAll(() => {
    givenProps.onInit = jest.fn();
    useDataContextSpy = jest.spyOn(DataContextModule, 'useDataContext').mockReturnValue({
      setIsLoading: jest.fn(),
    });
    usePlayContextSpy = jest.spyOn(PlayContextModule, 'usePlayContext').mockReturnValue({
      setCurrentTurn: jest.fn(),
      setScores: jest.fn(),
    });
    resetGameSpy = jest.spyOn(Api, 'resetGame').mockResolvedValue({ scores: null, currentTurn: null });
    modalInstance.close = jest.fn();
  });

  afterEach(() => {
    givenProps.onInit.mockClear();
    useDataContextSpy().setIsLoading.mockClear();
    useDataContextSpy.mockClear();
    usePlayContextSpy().setCurrentTurn.mockClear();
    usePlayContextSpy().setScores.mockClear();
    usePlayContextSpy.mockClear();
    resetGameSpy.mockClear();
    modalInstance.close.mockClear();
  });

  afterAll(() => {
    givenProps.onInit.mockRestore();
    useDataContextSpy().setIsLoading.mockRestore();
    useDataContextSpy.mockRestore();
    usePlayContextSpy().setCurrentTurn.mockRestore();
    usePlayContextSpy().setScores.mockRestore();
    usePlayContextSpy.mockRestore();
    resetGameSpy.mockRestore();
    modalInstance.close.mockRestore();
    jest.restoreAllMocks();
    jest.resetModules();
    jest.clearAllMocks();
  });

  it('should render the component correctly', () => {
    // Given / When
    const component = shallow(<ModalGameOver {...givenProps} />);

    // Then
    expect(component).toHaveLength(1);
    expect(component.find(Modal)).toHaveLength(1);
    expect(component.find(Modal).props().footer.type.displayName).toBe(Button.displayName);
    expect(component.find(Modal).props().footer.props.children).toBe('Play Again');
    expect(component.find(Modal)).toMatchSnapshot();
  });

  it('should handle the modal init', () => {
    // Given
    const component = shallow(<ModalGameOver {...givenProps} />);

    // When
    component.find(Modal).props().onInit(modalInstance);

    // Then
    expect(givenProps.onInit).toHaveBeenCalledTimes(1);
    expect(givenProps.onInit).toHaveBeenLastCalledWith(modalInstance);
  });

  it('should handle the play again button', async () => {
    // Given
    const component = await shallow(<ModalGameOver {...givenProps} />);

    // When
    await component.find(Modal).props().onInit(modalInstance);
    await component.find(Modal).props().footer.props.onClick();

    // Then
    expect(useDataContextSpy().setIsLoading).toHaveBeenCalledTimes(2);
    expect(useDataContextSpy().setIsLoading).toHaveBeenNthCalledWith(1, true);
    expect(useDataContextSpy().setIsLoading).toHaveBeenNthCalledWith(2, false);
    expect(resetGameSpy).toHaveBeenCalledTimes(1);
    expect(usePlayContextSpy().setCurrentTurn).toHaveBeenCalledTimes(1);
    expect(usePlayContextSpy().setCurrentTurn).toHaveBeenLastCalledWith(null);
    expect(usePlayContextSpy().setScores).toHaveBeenCalledTimes(1);
    expect(usePlayContextSpy().setScores).toHaveBeenLastCalledWith(null);
    expect(modalInstance.close).toHaveBeenCalledTimes(1);
  });
});
