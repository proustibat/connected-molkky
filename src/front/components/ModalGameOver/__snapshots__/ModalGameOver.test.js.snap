// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalGameOver should render the component correctly 1`] = `
<Modal
  footer={
    <Button
      className=""
      disabled={false}
      onClick={[Function]}
      size="btn-large"
      style={
        Object {
          "margin": 0,
        }
      }
    >
      Play Again
    </Button>
  }
  id="game-over-modal"
  onCloseEnd={null}
  onInit={[Function]}
  styleTitle={
    Object {
      "justifyContent": "center",
    }
  }
  title="GAME OVER"
>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "height": "calc(100% - 43px)",
        "justifyContent": "space-evenly",
        "margin": 0,
        "padding": 0,
      }
    }
  >
    <p
      className="flow-text"
      style={
        Object {
          "margin": 0,
          "textAlign": "center",
        }
      }
    >
      Boo-ya!
      <br />
      Team Cat
       
      wins!
    </p>
    <div
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <TrophySVG
        height="48"
        style={
          Object {
            "display": "block",
            "fill": "#143746",
            "height": "15rem",
            "margin": "auto",
            "width": "15rem",
          }
        }
        viewBox="0 0 48 48"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      />
      <CatSVG
        style={
          Object {
            "left": "calc(50% - 2.25rem)",
            "marginLeft": "0.6rem",
            "marginRight": "0.6rem",
            "position": "absolute",
            "top": "5rem",
            "width": "3.5rem",
          }
        }
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      />
      <p
        style={
          Object {
            "color": "#e2e2e2",
            "fontSize": "2rem",
            "left": "calc(50% - 2rem)",
            "position": "absolute",
            "textAlign": "center",
            "top": 0,
            "width": "4rem",
          }
        }
      >
        50
      </p>
    </div>
    <p
      className="flow-text"
      style={
        Object {
          "margin": 0,
          "textAlign": "center",
        }
      }
    >
      Bwah-hah-hah!
      <br />
      <span
        aria-label="thumb down"
        role="img"
      >
        ðŸ‘Ž
      </span>
      <DogSVG
        style={
          Object {
            "verticalAlign": "middle",
            "width": "2rem",
          }
        }
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      />
      <br />
      Team Dog
       
      loses with
       
      35
      !
    </p>
  </div>
</Modal>
`;
